# The Useful Five

In this notebook I have discussed about 5 useful functions that every numpy user must know.

## What is Numpy ?

Numpy is a numerical computaion library for python that supports large, multi-dimensional arrays and matrices, along with a large collection of high-level mathematical functions to operate on these arrays. It is one of the core data science libraries.

In [1]:
!pip install jovian --upgrade -q

In [2]:
import jovian

In [3]:
jovian.commit(project='numpy-array-operations')

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Please enter your API key ( from https://jovian.ml/ ):
API KEY: ········
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

Let's begin by importing Numpy and listing out the functions covered in this notebook.

In [4]:
import numpy as np

## Functions Explained here

In [88]:
# List of functions explained 
function1 = np.unravel_index  
function2 = np.add
function3 = np.reshape
function4 = np.split
function5 = np.cross

## Function 1 - np.unravel_index

This function is used to find the position of the nth element if we provide the shape of a  multidimensional array

In [61]:
# Example 1 - working 
shape_of_array = (5,4,6)
position = np.unravel_index(70, shape_of_array)
print(position)

(2, 3, 4)


It provides the position of the 70th element of an array of shape (5,4,6)

In [62]:
# Example 2 - working
shape_of_array = (5,4,6)
position = np.unravel_index(119, shape_of_array)
print(position)

(4, 3, 5)


It prints the position of the 119th element of an array of shape (5,4,6)

In [63]:
# Example 3 - breaking (to illustrate when it breaks)
shape_of_array = (5,4,6)
position = np.unravel_index(156, shape_of_array)
print(position)

ValueError: index 156 is out of bounds for array with size 120

Above example breaks because we demanded the position of an element that exceeds the size of our array.We demanded the position of 156th element but the max we can go is 119 as the length of the array is 120 in this case. It can be fixed by choosing a position between inclusive range of 0 to 119

A very handy function when you want to find the position of certain element in a multidimensional array.

In [64]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

## Function 2 - np.add

Function to add the elements of two arrays

In [65]:
# Example 1 - working
A = np.array([[1,2,3], [1,2,3]])
B = np.array([[3,2,1],[3,2,1]])
result = np.add(A,B)
result

array([[4, 4, 4],
       [4, 4, 4]])

Adds the elements of two  multidimensional arrays

In [66]:
# Example 2 - working
C = np.array([1,2,3,4])
D = np.array([4,3,2,1])
result2 = np.add(C,D)
result2

array([5, 5, 5, 5])

Adds the elements of the two provided arrays

In [67]:
# Example 3 - breaking (to illustrate when it breaks)
E = np.array([1,2,3,4])
F = np.array([4,3,2,1, 0])
result3 = np.add(E,F)
result3

ValueError: operands could not be broadcast together with shapes (4,) (5,) 

We tried adding two arrays of different shapes hence we see this error it can be fixed by using arrays of the same shape.

This function is super useful for adding elements of two arrays of the same shape.

In [68]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

## Function 3 - np.reshape

Function to change the shape of a numpy array

In [69]:
# Example 1 - working
array1 = np.arange(1,7)
np.reshape(array1, (2,3))

array([[1, 2, 3],
       [4, 5, 6]])

The above example reshapes a (6,) shape array into a (2,3) shape array

In [70]:
# Example 2 - working
array2 = np.array([[1,2,3], [4,5,6],[7,8,9]])
np.reshape(array2, (9))

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

Here we reshape an (3,3) shape  array into a (9,) shape array

In [71]:
# Example 3 - breaking (to illustrate when it breaks)
array3 = np.arange(1,6)
np.reshape(array3,(2,5))

ValueError: cannot reshape array of size 5 into shape (2,5)

Our example breaks because we tried reshaping an array of length 5 into an array of shape(2,5) which is not possible we can fix it specifing the proper shape to reshape the given array to . 

Very handy function for changing the shape of a given array into our desired shape.

In [72]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

## Function 4 - np.split

This function can be used to split our array into n arrays of equal size where n is provided by the user.

In [75]:
# Example 1 - working
array_a = np.arange(6)
np.split(array_a, 3)

[array([0, 1]), array([2, 3]), array([4, 5])]

Here an array of shape (6,) is split into 3 sub-arrays of shape(2,)

In [81]:
# Example 2 - working
array_b = np.arange(12).reshape(2,6)
np.split(array_b, 2)

[array([[0, 1, 2, 3, 4, 5]]), array([[ 6,  7,  8,  9, 10, 11]])]

Here we split an array of shape (2,6) into 3 sub-arrays of shape(6,)

In [86]:
# Example 3 - breaking (to illustrate when it breaks)
array_c = np.arange(16).reshape(4,4)
np.split(array_c, 5)

ValueError: array split does not result in an equal division

Here we tried spliting an array of shape(4,4) into 5 sub-arrays of equal shape which was not possible we can fix this error by choosing a correct number to split our array into equal size an example of it maybe splitting it into 4 parts which is possible.

This function is very useful if we want to split a given array into some equal sub-arrays and these subarrays can be used to perform further operations.

In [87]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

## Function 5 - np.cross

This function is used to produce the cross product of two arrays or vectors

In [93]:
# Example 1 - working
first = np.arange(6).reshape(2,3)
second = np.arange(5,11).reshape(2,3)
np.cross(first,second)

array([[-5, 10, -5],
       [-5, 10, -5]])

Here we obtain the cross product of two arrays of shape(2,3) with the default axis =1 for both the arrays.

In [97]:
# Example 2 - working
first = np.arange(6).reshape(2,3)
second = np.arange(5,11).reshape(2,3)
np.cross(first,second,axisa=0,axisb=0)

array([-15, -15, -15])

We used the same arrays here but the axis have been changed hence we see a different result as the computation is performed along the 0 axis.

In [98]:
# Example 3 - breaking (to illustrate when it breaks)
first = np.arange(6).reshape(2,3)
second = np.arange(5,11).reshape(2,3)
np.cross(first,second,axisa=2,axisb=2)

AxisError: axisa: axis 2 is out of bounds for array of dimension 2

We again used the same array but this time with axis =2 which resulted in a out of bounds error . we can fix this error by choosing an axis that comes under the bounds of the the arrays like 0,1.

This function has high mathematical importance as it helps us in easily obtaining the cross product of two vectors.

In [100]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'

## Conclusion

In this notebook I discussed about five important numpy functions and there importance along with 2 working and one breaking example. Further we can explore more functions of numpy as there are many other useful features in the numpoy library.

## Reference Links
Provide links to your references and other interesting articles about Numpy arrays:
* Numpy official tutorial : https://numpy.org/doc/stable/user/quickstart.html
* W3schools Numpy tutorial:
https://www.w3schools.com/python/numpy_intro.asp#:~:text=NumPy%20is%20a%20python%20library,you%20can%20use%20it%20freely.
* Edureka Numpy tutorial: https://www.edureka.co/blog/python-numpy-tutorial/

In [102]:
jovian.commit()

<IPython.core.display.Javascript object>

[jovian] Attempting to save notebook..
[jovian] Updating notebook "ajetely/numpy-array-operations" on https://jovian.ml/
[jovian] Uploading notebook..
[jovian] Capturing environment..
[jovian] Committed successfully! https://jovian.ml/ajetely/numpy-array-operations


'https://jovian.ml/ajetely/numpy-array-operations'